### DISCUSSION TITLE
Problem with query including printf

### DISCUSSION THREAD

**Discussion by javafanboy** [duckdb/duckdb#14188] [Q&A]
I have a table with two columns one VARCHAR called message with messages that optionally may contain a %s and another called messageparameters that is an array of VARCHAR that is null if there are no %s in the message and otherwise holds the desired string to match with the %s.
I would like to create a query that finds all the rows where message contains a %s and for thise creates a third column in the result that uses a "printf" function to produce the full message. 
I have tried the following:
```
SELECT
      message,
      messageparameters,
      CASE
          WHEN messageparameters IS NOT NULL AND array_length(messageparameters) >= 1 THEN
              printf(message, messageparameters[1])  -- Accessing first element safely
          WHEN messageparameters IS NULL THEN
              NULL  -- Explicitly handle NULL case
          ELSE
              'No parameters available'  -- Handle cases with no parameters
      END AS full
  FROM ram_table
  WHERE message LIKE '%\%s%' ESCAPE '\';
```
  
 but it produces the error message "Invalid Input Error: Argument index "1" out of range.
  
This is just a first simple test, in reality I would have liked to handle strings with up to say ten %s and matching number of parameters in the array.

I would have been really nice if printf would accept an array column for the second argument (it does but the result is the whole array geting inserted in place of the %s) - as it is now I guess I need a case statement alternative for each array length from 0 to 5 or is there some smarter solution?

**Comment by javafanboy** [comment]
Found an error in the data where a few rows contained %s%s but only one element in the array explaining the first problem but I would still like sugestions on how to best handle messages with different number of %s (and matching number of parameters.... ie is there a better way than a case statement for each number of parameters like this:

```
SELECT
      message,
      messageparameters,
      CASE
          WHEN messageparameters IS NULL THEN message
          WHEN array_length(messageparameters) = 0 THEN message
          WHEN array_length(messageparameters) = 1 THEN printf(message, messageparameters[1])
          WHEN array_length(messageparameters) = 2 THEN printf(message, messageparameters[1], messageparameters[2])
          WHEN array_length(messageparameters) = 3 THEN printf(message, messageparameters[1], messageparameters[2], messageparameters[3])
          WHEN array_length(messageparameters) = 4 THEN printf(message, messageparameters[1], messageparameters[2], messageparameters[3], messageparameters[4])
          WHEN array_length(messageparameters) = 5 THEN printf(message, messageparameters[1], messageparameters[2], messageparameters[3], messageparameters[4], messageparameters[5])
          ELSE NULL  -- In case there are more than 5 parameters
      END AS fullmessage
  FROM ram_table;
```

**Comment by soerenwolfers** [comment]
You can include code blocks using triple back ticks followed by the language (or not, if your just writing pseudo code) as such:

````
```sql
SELECT 1
```
````

and it'll render beautifully as such: 

```sql
SELECT 1
```

**Comment by soerenwolfers** [comment]
Regarding your actual question:

1: Your error is likely due to the same phenomenon mentioned in https://github.com/duckdb/duckdb/issues/13466 and the two tickets linked there
2: As far as I know, there is no such solution. The keyword here is "argument unpacking", and while this was recently added for the `COLUMNS` expression, it doesn't seem possible for regular arguments.

**Comment by javafanboy** [comment]
I tried another solution to handle any number of arguments to printf taken from a columns that is an array of VARCHAR.

```
CASE 
    WHEN messageParameters IS NULL OR array_length(messageParameters) = 0 THEN message
    ELSE printf(message, VARIADIC messageParameters)
END AS formattedMessage

```

but it did not work either - the messageParameters array of varchars was treated as ONE argument and was expanded as its content with [] etcl. rather than getting flatended.
